### ISSUE TITLE
[Story] Check ESLint consistency between IDE and CLI output

### ISSUE THREAD

**Issue by tschaffter** [Sage-Bionetworks/sage-monorepo#2789]
### What product(s) is this story for?

Sage Monorepo

### Tasks

- [ ] Evaluate how much time `nx format:check` takes after identifying files that must be ignored by Prettier

### Have you linked this story to a GitHub Project?

- [X] I have linked this story to a GitHub Project and set its metadata.

**Comment by tschaffter** [comment]
## ESLint extension for VS Code

It's active by default but it's not specifically defined in VS Code config file, which we should do.

Here is information from the extension:

```
[Info  - 9:12:58 PM] ESLint server is starting.
[Info  - 9:13:00 PM] ESLint server running in node v20.14.0
[Info  - 9:13:00 PM] ESLint server is running.
[Info  - 9:13:01 PM] ESLint library loaded from: /workspaces/sage-monorepo/node_modules/.pnpm/eslint@8.57.0/node_modules/eslint/lib/api.js
```

Observation:

- The VS Code plugin is using a different version of Node.js (v20.14.0) than the Node.js version installed in the dev container (v20.17.0)
- The VS Code plugin runs `eslint@8.57.0` while the workspace installs `v8.57.0`.
  - TODO: try to downgrade the version installed in the workspace to see if the two versions always match. I expect that they are different version because of the different version of Node.js they run on.

**Comment by tschaffter** [comment]
## ESLint formatter feature

The extension can be used as a formatter, but it's not enable by default. For example, when opening a TypeScript file:

![image](https://github.com/user-attachments/assets/5b03ffce-b1e8-4692-bea9-a9d833465f1a)

To enable the formatter feature:

```json
"eslint.format.enable": true,
```

VS Code now shows:

![image](https://github.com/user-attachments/assets/9ac98da5-c0be-4914-aa6d-0283f16ea9a5)

**Comment by tschaffter** [comment]
## Format on Save

Currently, ESLint is formatting TS files on save, though this is something I would like to configure specifically.

The formatter feature of the ESLint extension controlled by `eslint.format.enable` does not affect formatting on save.

ESLint does not respect this config (it will still format on save):

```
  "[typescript]": {
    "editor.defaultFormatter": "dbaeumer.vscode-eslint",
    "editor.formatOnSave": false
  },
```

This section describes how to enable and disable auto fix: https://github.com/microsoft/vscode-eslint?tab=readme-ov-file#version-204

**Comment by tschaffter** [comment]
## Format JSON files with ESLint-Prettier

This ESLint configuration work to format the files `test.json` and `test/test.json`.

`.eslintrc.json`:

```json
{
  "root": true,
  "overrides": [
    {
      "files": ["*.json", "*.jsonc", "*.json5", ".vscode/test.json"],
      "parser": "jsonc-eslint-parser",
      "extends": ["plugin:prettier/recommended"],
      "rules": {}
    }
  ]
}
```

Example:

![image](https://github.com/user-attachments/assets/70d3eaab-5ad7-435d-b737-bba9844e1191)

Saving the file will format it thanks to the VS Code configuration we have previously identified.

### Issue

The file `.vscode/test.json`, which has the same content as the previous file, is not formatted for an unknown reason.

![image](https://github.com/user-attachments/assets/0f9a3ad4-9616-4e8f-89a6-d2cb04315eb9)

**Comment by tschaffter** [comment]
## `nx format:check` and `nx format:write`

This script formats all the files in the monorepo using Prettier.

Source: https://github.com/nrwl/nx/blob/master/packages/nx/src/command-line/format/format.ts

Example:

```console
$ time nx format:check --all
.devcontainer/devcontainer.json
.eslintrc.json
.github/PULL_REQUEST_TEMPLATE/default.md
.github/renovate.json
.github/workflows/build-docs.yml
.github/workflows/ci.yml
.github/workflows/lint-dockerfiles.yml
.github/workflows/lint-pr.yml
.github/workflows/schematic-api-ci.yml
.github/workflows/sonar-scan-pull-request.yml
.github/workflows/sonar-scan-push.yml
.github/workflows/update-oc-db-csv-files.yml
.prettierrc
.vscode/extensions.json
apps/agora/apex/project.json
apps/agora/api-docs/project.json
apps/agora/api/src/api.ts
apps/agora/api/src/components/dataversion.ts
apps/agora/api/src/models/dataversion.ts
apps/agora/api/webpack.config.js
apps/agora/app/src/app/app.config.ts
apps/agora/app/src/app/app.routes.ts
apps/agora/app/src/main.ts
apps/agora/data/.venv/lib/python3.11/site-packages/agora_data-0.1.0.dist-info/direct_url.json
apps/agora/data/.venv/lib/python3.11/site-packages/httpcore-1.0.5.dist-info/licenses/LICENSE.md
apps/agora/data/.venv/lib/python3.11/site-packages/httpx-0.27.0.dist-info/licenses/LICENSE.md
apps/agora/data/.venv/lib/python3.11/site-packages/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/ieee_754.md
apps/agora/data/README.md
apps/iatlas/api/README.md
apps/iatlas/data/.venv/lib/python3.10/site-packages/decorator-5.1.1.dist-info/pbr.json
apps/iatlas/data/.venv/lib/python3.10/site-packages/numpy/random/LICENSE.md
apps/iatlas/data/.venv/lib/python3.10/site-packages/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/ieee_754.md
apps/iatlas/data/README.md
apps/model-ad/apex/project.json
apps/model-ad/api-docs/project.json
apps/model-ad/api/project.json
apps/model-ad/api/src/main/resources/openapi.yaml
apps/model-ad/api/templates/config.yaml
apps/model-ad/app/server.ts
apps/model-ad/app/src/app/app.config.ts
apps/model-ad/app/src/app/app.routes.ts
apps/model-ad/app/src/index.html
apps/model-ad/app/src/main.ts
apps/model-ad/mongo/project.json
apps/openchallenges/apex/README.md
apps/openchallenges/api-docs/README.md
apps/openchallenges/app/e2e/home.spec.ts
apps/openchallenges/app/server.ts
apps/openchallenges/app/src/_app-theme.scss
apps/openchallenges/app/src/app/app.component.ts
apps/openchallenges/app/src/app/app.config.ts
apps/openchallenges/app/src/app/app.routes.ts
apps/openchallenges/app/src/app/google-tag-manager/google-tag-manager-id.provider.ts
apps/openchallenges/app/src/app/initialize-keycloak.factory.ts
apps/openchallenges/app/src/assets/silent-check-sso.html
apps/openchallenges/app/src/index.html
apps/openchallenges/app/src/main.ts
apps/openchallenges/auth-service/docker-compose.yml
apps/openchallenges/challenge-service/src/main/resources/db/README.md
apps/openchallenges/challenge-service/src/main/resources/openapi.yaml
apps/openchallenges/challenge-service/templates/config.yaml
apps/openchallenges/challenge-to-elasticsearch-service/docker-compose.yml
apps/openchallenges/config-server-repository/README.md
apps/openchallenges/config-server/README.md
apps/openchallenges/core-service/docker-compose.yml

real    1m1.142s
user    0m2.377s
sys     0m0.998s
```

Observations:
- It takes about 1 minutes to format all the files given the current content of the monorepo.
- There are files that we may not want to format, e.g. the pnpm lock file (see below).
  - Lock files are generally large and may take time to format for no benefit.
  - There are "machine" files like lock files that humans don't read where readability is not needed.
  - Formatted/prettyfied files may make the file larger (see example below)

![image](https://github.com/user-attachments/assets/b69401d0-226e-477a-bf80-262dfab2e789)

Formatted pnpm lock file is larger:

```
# Before formatting
vscode@1fb8313797af:/workspaces/sage-monorepo$ ls -al pnpm-lock.yaml 
-rw-r--r--. 1 vscode vscode 1229364 Sep  5 00:05 pnpm-lock.yaml

# After formatting
vscode@1fb8313797af:/workspaces/sage-monorepo$ ls -al pnpm-lock.yaml 
-rw-r--r--. 1 vscode vscode 1158059 Sep  5 14:51 pnpm-lock.yaml
```

**Comment by tschaffter** [comment]
## Formatting Vs Linting

Prettier is a formatter. ESLint is a linter.

According to this [thread](https://www.reddit.com/r/node/comments/w95n80/why_use_prettier_if_eslint_can_format/), ESLint tried to do both formatting and linting, but is now focusing on linting.

The above thread also mentions concerns about using ESLint to "auto fix" issues, in particular auto fixing some linting issues may be sub-optimal.

## Prettier vs. ESLint: A Comparison

Based on the above Reddit thread:

Feature | Prettier | ESLint
-- | -- | --
Primary Function | Code formatting | Code linting and formatting
Opinionated | Highly opinionated (enforces a specific style) | Less opinionated (allows customization)
Configurability | Limited configuration options | Extensive configuration options
Performance | Generally faster | Can be slower, especially with complex rulesets
Ease of Use | Simple to set up and use | Requires more configuration and setup
Focus | Formatting (spacing, quotes, etc.) | Linting (style guidelines, potential errors)
Compatibility | Compatible with most editors and linters | May conflict with certain formatting rules
Use Case | Ensuring consistent code formatting | Enforcing coding standards and detecting potential issues

In summary: Prettier is excellent for enforcing a consistent code style quickly and easily, while ESLint offers more flexibility and control for linting and style guidelines. Many developers use both tools together to achieve a balance between consistency and customization.
